File: crates/nexus-bootstrap/src/terminal.rs
//! Typed IPC-client helpers for the `com.nexus.terminal` core plugin.
//!
//! CLI / TUI callers reach the terminal subsystem exclusively through these
//! helpers — no direct `nexus-terminal` dependency, matching the pattern of
//! [`crate::storage`] and [`crate::database`].
//!
//! Each helper:
//!
//! 1. Serializes arguments to JSON.
//! 2. `await`s the async [`IpcInvoker::ipc_call`] on the provided invoker.
//! 3. Deserializes the response into a typed DTO.
//!
//! DTOs mirror the corresponding structs in `nexus_terminal` but contain
//! only the fields callers read. Extra fields in the response are ignored
//! by serde.
//!
//! BL-147 — helpers take `&dyn IpcInvoker` rather than `&Runtime`, so the
//! same surface works against both local and remote (`ssh://`) forges. Each
//! helper is `async`; sync callers wrap with `rt.block_on(...)`.

use std::time::Duration;

use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};

use crate::invoker::IpcInvoker;

const TERMINAL_PLUGIN: &str = "com.nexus.terminal";
const IPC_TIMEOUT: Duration = Duration::from_secs(30);

// ── DTOs ─────────────────────────────────────────────────────────────────────

/// Mirror of `nexus_terminal::OutputLine`. Timestamps are Unix-ms per the
/// crate's JSON encoding of `SystemTime`.
#[derive(Debug, Clone, Deserialize)]
pub struct OutputLine {
    /// Milliseconds since Unix epoch at first ingestion.
    pub timestamp_ms: u64,
    /// ANSI-stripped text content (no trailing newline).
    pub content: String,
    /// Raw bytes including ANSI sequences.
    #[serde(default)]
    pub raw: Vec<u8>,
    /// Adjacent-repeat counter (1 for a distinct line).
    #[serde(default = "one")]
    pub repeats: u32,
}

const fn one() -> u32 {
    1
}

/// Mirror of `nexus_terminal::SessionInfo`.
#[derive(Debug, Clone, Deserialize)]
pub struct SessionInfo {
    /// Opaque session id.
    pub id: String,
    /// Human-readable label.
    pub name: String,
    /// Shell path (empty when unknown to the server).
    #[serde(default)]
    pub shell: String,
    /// Working directory, if known.
    #[serde(default)]
    pub working_dir: Option<String>,
    /// Current line count in the buffer.
    #[serde(default)]
    pub line_count: usize,
    /// Unix-seconds creation timestamp.
    #[serde(default)]
    pub created_at: u64,
}

/// Arguments for the `create_session` helper. Matches
/// `nexus_terminal::CreateSessionArgs`.
#[derive(Debug, Clone, Default, Serialize)]
pub struct CreateSessionArgs {
    /// Optional human-readable label.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    /// Absolute path to the shell executable. Falls back to detection
    /// when omitted.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub shell: Option<String>,
    /// Extra args for the shell binary (after the program name).
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub shell_args: Vec<String>,
    /// Working directory.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub working_dir: Option<String>,
    /// Extra env vars.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub env: Vec<(String, String)>,
}

#[derive(Debug, Deserialize)]
struct CreateSessionResponse {
    id: String,
}

#[derive(Debug, Deserialize)]
struct PumpResponse {
    bytes: usize,
}

// ── Internal helper ──────────────────────────────────────────────────────────

async fn call<T: serde::de::DeserializeOwned>(
    invoker: &(dyn IpcInvoker + Send + Sync),
    command: &str,
    args: serde_json::Value,
) -> Result<T> {
    let value = invoker
        .ipc_call(TERMINAL_PLUGIN, command, args, IPC_TIMEOUT)
        .await
        .with_context(|| format!("terminal ipc call '{command}' failed"))?;
    serde_json::from_value(value)
        .with_context(|| format!("terminal ipc response '{command}' decode failed"))
}

// ── Public helpers ───────────────────────────────────────────────────────────

/// Spawn a new PTY session and return its id.
///
/// # Errors
/// Propagates any IPC / shell-spawn failure surfaced by the core plugin.
pub async fn create_session(
    invoker: &(dyn IpcInvoker + Send + Sync),
    args: CreateSessionArgs,
) -> Result<String> {
    let resp: CreateSessionResponse =
        call(invoker, "create_session", serde_json::to_value(args)?).await?;
    Ok(resp.id)
}

/// Graceful shutdown of the session's PTY child via the §5.1 signal
/// ladder.
///
/// # Errors
/// Propagates any IPC / signal-delivery failure.
pub async fn close_session(invoker: &(dyn IpcInvoker + Send + Sync), id: &str) -> Result<()> {
    let _: serde_json::Value = call(
        invoker,
        "close_session",
        serde_json::json!({ "id": id }),
    )
    .await?;
    Ok(())
}

/// Write `input` to the session's stdin, appending a newline if absent.
///
/// # Errors
/// Propagates any IPC / write failure.
pub async fn send_input(
    invoker: &(dyn IpcInvoker + Send + Sync),
    id: &str,
    input: &str,
) -> Result<()> {
    let _: serde_json::Value = call(
        invoker,
        "send_input",
        serde_json::json!({ "id": id, "input": input }),
    )
    .await?;
    Ok(())
}

/// Send raw bytes to the session's stdin (no newline added).
///
/// # Errors
/// Propagates any IPC / write failure.
pub async fn send_raw_input(
    invoker: &(dyn IpcInvoker + Send + Sync),
    id: &str,
    data: &[u8],
) -> Result<()> {
    let _: serde_json::Value = call(
        invoker,
        "send_raw_input",
        serde_json::json!({ "id": id, "data": data }),
    )
    .await?;
    Ok(())
}

/// Drain the PTY into the session's line buffer. Blocks up to
/// `timeout_ms` for the first byte. Returns the byte count drained.
///
/// # Errors
/// Propagates any IPC / read failure.
pub async fn pump(
    invoker: &(dyn IpcInvoker + Send + Sync),
    id: &str,
    timeout_ms: u64,
) -> Result<usize> {
    let resp: PumpResponse = call(
        invoker,
        "pump",
        serde_json::json!({ "id": id, "timeout_ms": timeout_ms }),
    )
    .await?;
    Ok(resp.bytes)
}

/// Read a window of structured lines from the session buffer.
///
/// # Errors
/// Propagates any IPC failure.
pub async fn read_output(
    invoker: &(dyn IpcInvoker + Send + Sync),
    id: &str,
    start: Option<usize>,
    count: Option<usize>,
) -> Result<Vec<OutputLine>> {
    let mut args = serde_json::json!({ "id": id });
    if let Some(s) = start {
        args["start"] = s.into();
    }
    if let Some(c) = count {
        args["count"] = c.into();
    }
    call(invoker, "read_output", args).await
}

/// List every session the server knows about.
///
/// # Errors
/// Propagates any IPC failure.
pub async fn list_sessions(
    invoker: &(dyn IpcInvoker + Send + Sync),
) -> Result<Vec<SessionInfo>> {
    call(invoker, "list_sessions", serde_json::json!({})).await
}

/// Look up metadata for one session.
///
/// # Errors
/// Propagates any IPC failure; returns a hard error (not `None`) for
/// unknown ids so callers can distinguish "gone" from "empty buffer".
pub async fn get_session_info(
    invoker: &(dyn IpcInvoker + Send + Sync),
    id: &str,
) -> Result<SessionInfo> {
    call(
        invoker,
        "get_session_info",
        serde_json::json!({ "id": id }),
    )
    .await
}

File: app/src/ipc/terminal.ts
// Typed wrappers for the terminal core plugin Tauri commands.
//
// These forward straight to `com.nexus.terminal` via kernel IPC (see
// `crates/nexus-app/src/terminal.rs`). Mirrors the helper module
// pattern in `ipc/editor.ts` and the sibling
// `nexus_bootstrap::terminal` Rust helper used by `nexus-tui` — the
// Tauri shell reaches the terminal engine through `invoke(…)` and
// never links `nexus-terminal` directly.

import { invoke } from "@tauri-apps/api/core";

/** Structured line emitted by `term_read_output`. */
export interface OutputLine {
  /** Milliseconds since Unix epoch at first ingestion. */
  timestamp_ms: number;
  /** ANSI-stripped text content (no trailing newline). */
  content: string;
  /** Raw bytes as received from the PTY (includes ANSI sequences). */
  raw: number[];
  /** Adjacent-repeat counter. */
  repeats: number;
}

/** Session metadata returned by `term_get_session_info` / `term_list_sessions`. */
export interface SessionInfo {
  id: string;
  name: string;
  shell: string;
  working_dir: string | null;
  line_count: number;
  created_at: number;
}

/** Arguments accepted by `term_create_session`; every field is optional. */
export interface CreateSessionArgs {
  name?: string;
  shell?: string;
  shellArgs?: string[];
  workingDir?: string;
  env?: [string, string][];
}

interface CreateSessionResponse {
  id: string;
}

interface PumpResponse {
  bytes: number;
}

/**
 * Spawn a new PTY-backed session. Returns the fresh session id.
 *
 * All fields on `args` are optional — omit them to accept the
 * platform-default shell, cwd, and environment.
 */
export async function termCreateSession(
  args: CreateSessionArgs = {},
): Promise<string> {
  const resp = await invoke<CreateSessionResponse>("term_create_session", {
    name: args.name,
    shell: args.shell,
    shellArgs: args.shellArgs,
    workingDir: args.workingDir,
    env: args.env,
  });
  return resp.id;
}

/** Gracefully shut down a session via the §5.1 signal ladder. */
export function termCloseSession(id: string): Promise<void> {
  return invoke<void>("term_close_session", { id });
}

/**
 * Flush `input` (plus a trailing newline if absent) into the session's
 * stdin. Shell-level behaviour only — use [`termSendRawInput`] for
 * control sequences.
 */
export function termSendInput(id: string, input: string): Promise<void> {
  return invoke<void>("term_send_input", { id, input });
}

/** Write raw bytes to the PTY. No newline is added. */
export function termSendRawInput(id: string, data: number[]): Promise<void> {
  return invoke<void>("term_send_raw_input", { id, data });
}

/**
 * Drain the PTY into the session's line buffer. Returns the byte count
 * drained in this pump. Frontend polling loops read this to decide
 * whether to re-fetch the line snapshot.
 */
export async function termPump(
  id: string,
  timeoutMs = 50,
): Promise<number> {
  const resp = await invoke<PumpResponse>("term_pump", { id, timeoutMs });
  return resp.bytes;
}

/** Fetch a range of lines from the session's buffer. */
export function termReadOutput(
  id: string,
  start?: number,
  count?: number,
): Promise<OutputLine[]> {
  return invoke<OutputLine[]>("term_read_output", { id, start, count });
}

/** Literal or regex search. Returns line indices into the current buffer. */
export function termSearchOutput(
  id: string,
  query: string,
  isRegex = false,
): Promise<number[]> {
  return invoke<number[]>("term_search_output", { id, query, isRegex });
}

/** Metadata for one session. */
export function termGetSessionInfo(id: string): Promise<SessionInfo> {
  return invoke<SessionInfo>("term_get_session_info", { id });
}

/** Every session the server knows about. */
export function termListSessions(): Promise<SessionInfo[]> {
  return invoke<SessionInfo[]>("term_list_sessions");
}

// ─── Saved commands (PRD-09 §14.1) ─────────────────────────────────────────
//
// JSON-compatible view of `nexus_terminal::SavedCommand`. Field names use
// snake_case to match the Rust struct's default serde layout — the
// frontend store converts to/from camelCase at the boundary.

export interface SavedCommandDto {
  slug: string;
  name: string;
  shell: string;
  shell_cmd: string;
  working_dir: string | null;
  env_vars: Record<string, string>;
  env_file: string | null;
  icon: string;
  auto_restart: boolean;
  auto_restart_delay_ms: number;
  memory_limit_mb: number | null;
  sidebar_order: number | null;
  pre_commands: string[];
  created_at: number;
  updated_at: number;
}

export function termSavedList(): Promise<SavedCommandDto[]> {
  return invoke<SavedCommandDto[]>("term_saved_list");
}

export function termSavedCreate(
  command: SavedCommandDto,
): Promise<SavedCommandDto> {
  return invoke<SavedCommandDto>("term_saved_create", { command });
}

export function termSavedUpdate(
  command: SavedCommandDto,
): Promise<SavedCommandDto> {
  return invoke<SavedCommandDto>("term_saved_update", { command });
}

export function termSavedDelete(slug: string): Promise<void> {
  return invoke<void>("term_saved_delete", { slug });
}

export function termSavedReorder(
  slug: string,
  sidebarOrder: number | null,
): Promise<void> {
  return invoke<void>("term_saved_reorder", { slug, sidebarOrder });
}
